homeassistant/components/netatmo/__init__.py

```diff
@@ -61,9 +106,88 @@ def setup(hass, config):
+            EVENT_HOMEASSISTANT_STOP, dropwebhook)
+
+    def _service_addwebhook(service):
+        """Service to (re)add webhooks during runtime."""
+        url = service.data.get(CONF_URL)
+        if url is None:
+            url = NETATMO_WEBHOOK_URL
+        _LOGGER.info("Adding webhook for URL: %s", url)
+        NETATMO_AUTH.addwebhook(url)
+
+    hass.services.register(
+        DOMAIN, SERVICE_ADDWEBHOOK, _service_addwebhook,
+        schema=SCHEMA_SERVICE_ADDWEBHOOK)
+
+    def _service_dropwebhook(service):
+        """Service to drop webhooks during runtime."""
+        _LOGGER.info("Dropping webhook")
+        NETATMO_AUTH.dropwebhook()
+
+    hass.services.register(
+        DOMAIN, SERVICE_DROPWEBHOOK, _service_dropwebhook,
+        schema=SCHEMA_SERVICE_DROPWEBHOOK)
+
     return True
 
 
+def dropwebhook(hass):
+    """Drop the webhook subscription."""
+    NETATMO_AUTH.dropwebhook()
+
+
+async def handle_webhook(hass, webhook_id, request):
+    """Handle webhook callback."""
+    body = await request.text()
+    try:
+        data = json.loads(body) if body else {}
+    except ValueError:
+        return None
+
+    _LOGGER.debug("Got webhook data: %s", data)
+    published_data = {
+        ATTR_EVENT_TYPE: data.get(ATTR_EVENT_TYPE),
+        ATTR_HOME_NAME: data.get(ATTR_HOME_NAME),
+        ATTR_CAMERA_ID: data.get(ATTR_CAMERA_ID),
+        ATTR_MESSAGE: data.get(ATTR_MESSAGE)
+    }
+    if data.get(ATTR_EVENT_TYPE) == EVENT_PERSON:
+        for person in data[ATTR_PERSONS]:
+            published_data[ATTR_ID] = person.get(ATTR_ID)
+            published_data[ATTR_NAME] = NETATMO_PERSONS.get(
+                published_data[ATTR_ID], DEFAULT_PERSON)
+            published_data[ATTR_IS_KNOWN] = person.get(ATTR_IS_KNOWN)
+            published_data[ATTR_FACE_URL] = person.get(ATTR_FACE_URL)
+            hass.bus.async_fire(EVENT_BUS_PERSON, published_data)
+    elif data.get(ATTR_EVENT_TYPE) == EVENT_MOVEMENT:
+        published_data[ATTR_VIGNETTE_URL] = person.get(ATTR_VIGNETTE_URL)
+        published_data[ATTR_SNAPSHOT_URL] = person.get(ATTR_SNAPSHOT_URL)
+        hass.bus.async_fire(EVENT_BUS_MOVEMENT, published_data)
+    elif data.get(ATTR_EVENT_TYPE) == EVENT_HUMAN:
+        published_data[ATTR_VIGNETTE_URL] = person.get(ATTR_VIGNETTE_URL)
+        published_data[ATTR_SNAPSHOT_URL] = person.get(ATTR_SNAPSHOT_URL)
+        hass.bus.async_fire(EVENT_BUS_HUMAN, published_data)
+    elif data.get(ATTR_EVENT_TYPE) == EVENT_ANIMAL:
+        published_data[ATTR_VIGNETTE_URL] = person.get(ATTR_VIGNETTE_URL)
+        published_data[ATTR_SNAPSHOT_URL] = person.get(ATTR_SNAPSHOT_URL)
+        hass.bus.async_fire(EVENT_BUS_ANIMAL, published_data)
+    elif data.get(ATTR_EVENT_TYPE) == EVENT_VEHICLE:
+        hass.bus.async_fire(EVENT_BUS_VEHICLE, published_data)
+        published_data[ATTR_VIGNETTE_URL] = person.get(ATTR_VIGNETTE_URL)
+        published_data[ATTR_SNAPSHOT_URL] = person.get(ATTR_SNAPSHOT_URL)
```

Comment:
`person.get` should be `data.get` in all `elif` sections:

```Traceback (most recent call last):
  File "/Users/josh/.pyenv/versions/3.6.2/lib/python3.6/site-packages/homeassistant/components/webhook.py", line 78, in async_handle_webhook
    response = await webhook['handler'](hass, webhook_id, request)
  File "/Users/josh/.homeassistant/custom_components/netatmo/__init__.py", line 176, in handle_webhook
    published_data[ATTR_VIGNETTE_URL] = data.get(ATTR_VIGNETTE_URL)
UnboundLocalError: local variable 'person' referenced before assignment```

Reply:
Whoops. Copy & paster. 😆